lsp-mode.el

```diff
@@ -3581,8 +3591,36 @@ A reference is highlighted only if it is visible in a window."
                 'def-params td-params)))
 
 (defun lsp--get-document-symbols ()
-  (lsp-request "textDocument/documentSymbol"
-               `(:textDocument ,(lsp--text-document-identifier))))
+  "Get document symbols.
+
+If the buffer has not been modified since symbols were last
+retrieved, simply return the latest result.
+
+Else, if the request was initiated by Imenu updating its menu-bar
+entry, perform it asynchronously; i.e., give Imenu the latest
+result and then force a refresh when a new one is available.
+
+Else (e.g., due to intereactive use of `imenu' or `xref'),
+perform the request synchronously."
+  (if (= (buffer-chars-modified-tick) lsp--document-symbols-tick)
+      lsp--document-symbols
+    (let ((method "textDocument/documentSymbol")
+	  (params `(:textDocument ,(lsp--text-document-identifier))))
+      (if (not lsp--document-symbols-request-async)
+	  (setq lsp--document-symbols-tick (buffer-chars-modified-tick)
+		lsp--document-symbols (lsp-request method params))
+	(setq lsp--document-symbols-tick (buffer-chars-modified-tick))
```

Comment:
It should be like:
```
(let ( (tick (buffe-modi...)
         (buff (current-buffer))
  (.... 
     (lambda (result) 
         (with-current-buffer buff
               (setq lsp--document-symbol-tick tick)
                ....))
```
otherwise, the stale results might be used as a real one.

Reply:
But doesn't `:mode 'alive` already capture the buffer?

Reply:
Scratch that.  I see what you mean.

Reply:
Actually, you are right, buffer part is not needed, only the tick part.